design/lifecycle.md

```diff
@@ -0,0 +1,195 @@
+
+### Spark Job proposal
+
+The user can run a spark job that is in charge of enforcing the lifecycle policy.
+
+The spark job will delete all the data (addresses) of objects matching the
+lifecycle rules except for data that is referenced by other objects (objects not
+matching the lifecycle rule). In order to find the matching objects efficiently
+(without going over the history of all commits every time) the spark job will
+conclude the Date lifecycle job ran last time. Using that Date, the job will
+find only commits that might have objects matching the lifecycle policy and
+didn't match in the last run.
+
+   
+
+This job contains five main parts:
+1. Extract data from policy
+2. Find potential commits, and commit branches
+3. Find addresses that should be removed by lifecycle policy
+4. Find deduplications
+5. Delete all addresses found by lifecycle policy that have no deduplication
+
+each step will use the information generated in the previous step 
+
+#### Prepare data using rules
+ 
+The first step in the spark job would be to transform the current policy into `policy table`.
+The `policy table` will hold the policy for all run ids.
+The main reason for this is that the policy could constantly be changed, we need some way to remember what was the policy for each rule.
+Once we already have it we would use it to generate information needed for the next step to find the potential commits
+and will be used along the steps for validating if the policy is matched.
+
+The `policy table` will contain as follows:
+- run_id:  a unique id generated for this run
+- rule_id: an id provided by the user - has no influence but, an ID provided by the user.
+- prefix:  the prefix defined in the rule
+- branch:  
+- date_to_be_deleted: the date all the objects containing the prefix and exist in the branch should be deleted 
+
+For example the policy
+```json
+{
+   "rule1" : {
+   "prefix": "foo/bar",
+   "days": 10,
+   "enabled": true,
+   "branch_days": {
+   "b1": 5,
+   "b2": 8 
+    }
+   },
+   "rule2" : {
+   "prefix": "foo/zoo",
+   "enabled": true,
+   "branch_days": {
+      "b1": 5
+   }
+   }
+}
+```
+
+
+assuming today is 20/01/1998
```

Comment:
An I18N aside: please avoid date formats such as "DD/MM/YYYY".  They are very ill-defined, e.g. USAians often use "MM/DD/YYYY", and Europeans often use "DD.MM.YYYY". 

The world should have only one date format, and that format is ISO8601-style: "YYYY-MM-DD".  Apart from there being only one way to read it, it is also ASCII-sortable.